language/move-lang/src/lib.rs

```diff
@@ -354,33 +375,170 @@ fn verify_string(fname: &'static str, string: &str) -> Result<(), Error> {
                  are permitted.",
                 chr
             );
-            Err(vec![(loc, msg)])
+            Err(vec![vec![(loc, msg)]])
         }
     }
 }
 
-fn strip_comments(source: &str) -> String {
+/// Types to represent comments.
+pub type CommentMap = BTreeMap<&'static str, MatchedFileCommentMap>;
+pub type MatchedFileCommentMap = BTreeMap<ByteIndex, String>;
+pub type FileCommentMap = BTreeMap<Span, String>;
+
+/// Strips line and block comments from input source, and collects documentation comments,
+/// putting them into a map indexed by the span of the comment region. Comments in the original
+/// source will be replaced by spaces, such that positions of source items stay unchanged.
+/// Block comments can be nested.
+///
+/// Documentation comments are comments which start with
+/// `///` or `/**`, but not `////` or `/***`. The actually comment delimiters
+/// (`/// .. <newline>` and `/** .. */`) will be not included in extracted comment string. The
+/// span in the returned map, however, covers the whole region of the comment, including the
+/// delimiters.
+fn strip_comments(fname: &'static str, input: &str) -> Result<(String, FileCommentMap), Errors> {
```

Comment:
This looks great! But, in the long term I'd like to move all of this into the parser, in the hopes of later developing a full fidelity parser AST. 

Any thoughts on how much easier or harder it would be to move comments into the parser?

Reply:
As to the design: when I started looking at it, I was similar inclined of doing it in the parser. But then I started liking the current design (which extends what you had been doing until now) more and more. Some reasons:

- Very small amount of code compared to treating them as tokens. 
- Much better error messages (at least with current absence of recovery for parsing error). A misplaced comment does not led to a parsing error with subsequent bail out.
- No unnecessary storage in AST (through the N phases of the compiler) if you don't need them. Efficient access by just using the start position of a syntax item and accessing a table.

As to implementing it in the parser, this would be significant more work. Unless you just mean moving the strip-comments function into the lexer, instead of making comments become tokens. (So just some code moving.)

Reply:
No worries then. A problem for another day

For context, long term I'd like to see other tools like an auto-formatter. And those can be tricky if the comments aren't in the parer's AST

Reply:
I hope the auto-formatter isn't too fare out, I feel it is actually a quite important early step  to establish a language nowadays. I guess it would work well with the current model because while it would consume the AST (and attributes attached to it like documentation comments), it will transform it into a different domain, likely some kind of hierarchical text as used e.g. by Wadler style pretty printers. (https://crates.io/crates/pretty).